cloud/services/container/clusters/reconcile.go

```diff
@@ -0,0 +1,288 @@
+	log.Info("Reconciling cluster resources")
+
+	cluster, err := s.describeCluster(ctx)
+	if err != nil {
+		s.scope.GCPManagedControlPlane.Status.Ready = false
+		return ctrl.Result{}, err
+	}
+	if cluster == nil {
+		log.Info("Cluster not found, creating")
+		s.scope.GCPManagedControlPlane.Status.Ready = false
+		if err = s.createCluster(ctx); err != nil {
+			conditions.MarkFalse(s.scope.ConditionSetter(), infrav1exp.GKEControlPlaneReadyCondition, infrav1exp.GKEControlPlaneReconciliationFailedReason, clusterv1.ConditionSeverityError, err.Error())
+			conditions.MarkFalse(s.scope.ConditionSetter(), infrav1exp.GKEControlPlaneCreatingCondition, infrav1exp.GKEControlPlaneReconciliationFailedReason, clusterv1.ConditionSeverityError, err.Error())
+			return ctrl.Result{}, err
+		}
+		log.Info("Cluster provisioning in progress")
+		conditions.MarkFalse(s.scope.ConditionSetter(), infrav1exp.GKEControlPlaneReadyCondition, infrav1exp.GKEControlPlaneCreatingReason, clusterv1.ConditionSeverityInfo, "")
+		conditions.MarkTrue(s.scope.ConditionSetter(), infrav1exp.GKEControlPlaneCreatingCondition)
+		return ctrl.Result{RequeueAfter: 5 * time.Minute}, nil
+	}
+
+	switch cluster.Status {
+	case containerpb.Cluster_PROVISIONING:
+		log.Info("Cluster provisioning in progress")
+		conditions.MarkFalse(s.scope.ConditionSetter(), infrav1exp.GKEControlPlaneReadyCondition, infrav1exp.GKEControlPlaneCreatingReason, clusterv1.ConditionSeverityInfo, "")
+		conditions.MarkTrue(s.scope.ConditionSetter(), infrav1exp.GKEControlPlaneCreatingCondition)
+		s.scope.GCPManagedControlPlane.Status.Ready = false
+		return ctrl.Result{RequeueAfter: 5 * time.Minute}, nil
+	case containerpb.Cluster_RECONCILING:
+		log.Info("Cluster reconciling in progress")
+		conditions.MarkTrue(s.scope.ConditionSetter(), infrav1exp.GKEControlPlaneUpdatingCondition)
+		s.scope.GCPManagedControlPlane.Status.Ready = true
+		return ctrl.Result{}, nil
+	case containerpb.Cluster_STOPPING:
+		log.Info("Cluster stopping in progress")
+		conditions.MarkFalse(s.scope.ConditionSetter(), infrav1exp.GKEControlPlaneReadyCondition, infrav1exp.GKEControlPlaneDeletingReason, clusterv1.ConditionSeverityInfo, "")
+		conditions.MarkTrue(s.scope.ConditionSetter(), infrav1exp.GKEControlPlaneDeletingCondition)
+		s.scope.GCPManagedControlPlane.Status.Ready = false
+		return ctrl.Result{}, nil
+	case containerpb.Cluster_ERROR, containerpb.Cluster_DEGRADED:
+		var msg string
+		if len(cluster.Conditions) > 0 {
+			msg = cluster.Conditions[0].GetMessage()
+		}
+		log.Error(errors.New("Cluster in error/degraded state"), msg, "name", s.scope.GCPManagedControlPlane.Name)
+		conditions.MarkFalse(s.scope.ConditionSetter(), infrav1exp.GKEControlPlaneReadyCondition, infrav1exp.GKEControlPlaneErrorReason, clusterv1.ConditionSeverityError, "")
+		s.scope.GCPManagedControlPlane.Status.Ready = false
+		return ctrl.Result{}, nil
+	case containerpb.Cluster_RUNNING:
+		log.Info("Cluster running")
```

Comment:
Should we also set the `GKEControlPlaneUpdatingCondition` condition to false here explicitly? Or perhaps in the else of the `if needUpdate {`